crates/flowctl/src/raw/suggest_schema.rs

```diff
@@ -0,0 +1,205 @@
+use anyhow::Context;
+use bytes::Bytes;
+use doc::{inference::Shape, SchemaIndexBuilder, FailedValidation};
+use futures::{TryStreamExt, StreamExt};
+use journal_client::{broker, read::uncommitted::{ReadUntil, ReadStart, JournalRead, ExponentialBackoff, Reader}, list::list_journals, AuthHeader};
+use json::schema::{build::build_schema, types};
+use models::{
+    Capture, CaptureBinding, CaptureDef, CaptureEndpoint, Catalog, Collection, CollectionDef,
+    CompositeKey, ConnectorConfig, JsonPointer, Schema, ShardTemplate,
+};
+use proto_flow::{
+    capture::{request, Request},
+    flow::capture_spec::ConnectorType,
+    ops::Log
+};
+use bytelines::AsyncByteLines;
+use proto_grpc::broker::journal_client::JournalClient;
+use schema_inference::{json_decoder::JsonCodec, inference::infer_shape, shape, server::InferenceError, schema::SchemaBuilder};
+use serde_json::{json, value::RawValue, Value};
+use tokio::io::{BufReader, AsyncReadExt};
+use tokio_util::{compat::FuturesAsyncReadCompatExt, io::{ReaderStream, StreamReader}, codec::FramedRead};
+use tonic::{codegen::InterceptedService, transport::Channel};
+use std::{collections::BTreeMap, io::ErrorKind, time::{Duration, Instant}};
+use url::Url;
+
+use crate::{connector::docker_run, catalog::{fetch_live_specs, List, SpecTypeSelector, NameSelector, collect_specs}, dataplane::{self, fetch_data_plane_access_token}, collection::{CollectionJournalSelector, Partition, read::{journal_reader, ReadArgs}}};
+use crate::local_specs;
+use anyhow::anyhow;
+
+/// With some of our captures, we have an existing document schema for their collections, but we
+/// frequently run into issues with these document schemas: they are sometimes completely wrong
+/// about type of a field, or sometimes they are too narrow about the type (e.g. a value is marked
+/// as "integer", but it is actually "number"), other times they are missing some fields that are
+/// being captured into the collection.
+///
+/// This tool is built for the purpose of helping with updating the schema of these collections, by
+/// reading data from a collection, and the corresponding task ops logs (specifically document schema
+/// violation logs), and running schema inference on all documents of the collection as well as the
+/// documents that violated the existing schema, to come up with a new schema that will allow those
+/// documents to pass validation. The schema inference run starts with the existing task schema as
+/// its starting point, and widens that schema to allow the invalid documents to pass validation.
+///
+/// The reason for including all documents from the collection is that the task's existing schema
```

Comment:
@psFried the documents will have passed validation but that's because these collections all have `additionalProperties: true`, so missing fields are prominent among them.

Reply:
With `ReadArgs`, by setting `since=now` you can skip this at the moment